nx/lib/nx/random.ex

```diff
@@ -0,0 +1,419 @@
+  
+  To summarize, among other requirements, Nx's PRNG aims to:
+  
+  1. Ensure reproducibility;
+  
+  2. Parallelize well, both in terms of vectorization (generating array values)
+     and multi-replica, multi-core computation. In particular it should not use
+     sequencing constraints between random function calls.
+  """
+
+  import Nx.Defn, only: [deftransformp: 2, defn: 2, defnp: 2]
+
+  @nbits 32
+
+  @doc """
+  Create a pseudo-random number generator (PRNG) key given an integer seed.
+  
+  ## Examples
+  
+      iex> Nx.Random.key(12)
+      #Nx.Tensor<
+        u32[2]
+        [0, 12]
+      >
+  
+      iex> Nx.Random.key(999999999999)
+      #Nx.Tensor<
+        u32[2]
+        [232, 3567587327]
+      >
+  """
+  defn key(seed) do
+    k1 = Nx.right_shift(seed, 32)
+    k2 = Nx.bitwise_and(seed, 0xFFFFFFFF)
+
+    Nx.stack([k1, k2])
+    |> Nx.as_type({:u, 32})
+  end
+
+  @doc """
+  Splits a PRNG key into `num` new keys by adding a leading axis.
+  
+    ## Examples
+  
+      iex> key = Nx.Random.key(1701)
+      iex> Nx.Random.split(key, 2)
+      #Nx.Tensor<
+        u32[2][2]
+        [
+          [56197195, 1801093307],
+          [961309823, 1704866707]
+        ]
+      >
+  
+      iex> key = Nx.Random.key(999999999999)
+      iex> Nx.Random.split(key, 4)
+      #Nx.Tensor<
+        u32[4][2]
+        [
+          [3959978897, 4079927650],
+          [3769699049, 3585271160],
+          [3182829676, 333122445],
+          [3185556048, 1258545461]
+        ]
+      >
+  """
+  defn split(key, num \\ 2) do
+    assert_key(key)
+
+    impl(key, Nx.iota({num, 2}))
+  end
+
+  defn fold_in(key, data) do
+    assert_key(key)
+
+    impl(key, key(data))
+  end
+
+  deftransformp max_count_size(shape, bit_width) do
+    [
+      shape:
+        {rem(Nx.size(shape) * bit_width, 32)
+         |> Nx.any()
+         |> Nx.add(Nx.size(shape) * bit_width / 32)}
+    ]
+  end
+
+  defn random_bits(key, opts \\ []) do
+    assert_key(key)
+    opts = keyword!(opts, shape: {1}, bit_width: 32)
+    shape = opts[:shape]
+    bit_width = opts[:bit_width]
+
+    bits =
+      case bit_width do
+        64 ->
+          bits =
+            impl(key, Nx.concatenate([Nx.iota(shape), Nx.iota(shape) + Nx.size(shape)]))
+            |> Nx.reshape({2, :auto})
+            |> Nx.as_type({:u, bit_width})
+
+          bits[0] <<< 32 ||| bits[1]
+
+        32 ->
+          impl(key, Nx.iota(shape))
+
+        _ ->
+          impl(key, Nx.iota(shape))
+          |> Nx.as_type({:u, bit_width})
+      end
+
+    Nx.reshape(bits, shape)
+  end
+
+  defnp impl(key, count) do
+    shape = Nx.shape(count)
+
+    reshaped_key = Nx.reshape(key, {2, 1})
+
+    reshaped_count =
+      Nx.reshape(count, {:auto})
+      |> Nx.as_type({:u, 32})
+
+    threefry2x32(reshaped_key, reshaped_count)
+    |> Nx.reshape(shape)
+    |> Nx.as_type({:u, 32})
+  end
+
+  # Check count
+  defn threefry2x32(key, count) do
```

Comment:
Should this be private or documented?

Reply:
Should be private but I will leave tests commented for now.